gate-oauth2/src/main/java/com/netflix/spinnaker/gate/security/oauth2/SpinnakerUserInfoTokenServices.java

```diff
@@ -0,0 +1,304 @@
+      return String.valueOf(value).matches(mutateRegexPattern(requiredVal));
+    }
+
+    return value.equals(requiredVal);
+  }
+
+  public boolean hasAllUserInfoRequirements(Map<String, Object> details) {
+    if (userInfoRequirements == null || userInfoRequirements.isEmpty()) {
+      return true;
+    }
+
+    Map<String, String> invalidFields =
+        userInfoRequirements.entrySet().stream()
+            .filter(
+                entry -> {
+                  String reqKey = entry.getKey();
+                  String reqVal = entry.getValue();
+                  Object value = details.get(reqKey);
+
+                  if (value instanceof Collection<?> collection) {
+                    return collection.stream()
+                        .noneMatch(item -> valueMatchesConstraint(item, reqVal));
+                  }
+                  return !valueMatchesConstraint(value, reqVal);
+                })
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    if (!invalidFields.isEmpty() && log.isDebugEnabled()) {
+      log.debug(
+          "Invalid userInfo response: "
+              + invalidFields.entrySet().stream()
+                  .map(
+                      entry ->
+                          "got "
+                              + entry.getKey()
+                              + "="
+                              + details.get(entry.getKey())
+                              + ", wanted "
+                              + entry.getValue())
+                  .collect(Collectors.joining(", ")));
+    }
+
+    return invalidFields.isEmpty();
+  }
+
+  public static boolean isRegexExpression(String val) {
+    if (val.startsWith("/") && val.endsWith("/")) {
+      try {
+        Pattern.compile(val);
+        return true;
+      } catch (PatternSyntaxException ignored) {
+        return false;
+      }
+    }
+
+    return false;
+  }
+
+  public static String mutateRegexPattern(String val) {
+    // "/expr/" -> "expr"
+    return val.substring(1, val.length() - 1);
+  }
+
+  protected List<String> getRoles(Map<String, Object> details) {
+    if (userInfoMapping == null || userInfoMapping.getRoles() == null) {
+      return List.of();
+    }
+
+    Object roles = details.getOrDefault(userInfoMapping.getRoles(), List.of());
+
+    if (roles instanceof Collection<?> collection) {
+      // Some providers (Azure AD) return roles in this format: ["[\"role-1\", \"role-2\"]"]
+      if (!collection.isEmpty()
+          && collection.iterator().next() instanceof String firstRole
+          && firstRole.startsWith("[")) {
+        return parseJsonRoles(firstRole);
+      }
+      return collection.stream().map(Object::toString).collect(Collectors.toList());
+    }
+
+    if (roles instanceof String roleString) {
+      if (roleString.trim().isEmpty()) {
+        return new ArrayList<>();
+      }
+
+      return Arrays.asList(roleString.split("[, ]+"));
+    }
+
+    return List.of();
```

Comment:
The groovy code has
```
log.warn("unsupported roles value in details, type: ${roles.class}, value: ${roles}")
```
here.

Reply:
added the log statement

Reply:
https://github.com/spinnaker/gate/pull/1871/commits/e3cf19588f7fc395902161b55da92537a4c68334